core/trino-main/src/test/java/io/trino/exchange/TestLazyExchangeDataSource.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.exchange;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import io.trino.memory.context.SimpleLocalMemoryContext;
+import io.trino.metadata.ExchangeHandleResolver;
+import io.trino.operator.RetryPolicy;
+import io.trino.spi.QueryId;
+import io.trino.spi.exchange.ExchangeId;
+import org.testng.annotations.Test;
+
+import static io.trino.memory.context.AggregatedMemoryContext.newSimpleAggregatedMemoryContext;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestLazyExchangeDataSource
+{
+    @Test
+    public void testIsBlockedNonCancellable()
```

Comment:
It is more about cancelation not propagating from futures returned between separate calls done to `isBlocked`, each future is still cancellable. Also it only tests the "intialization" phase of LazyExchangeDataSource. Behaviour later in the lifecycle depends on how delegate `ExchangeDataSource` behaves.

Maybe more explanatory name of this test would be `testIsBlockedCancellationIsolationInInItializationPhase`
Kinda ugly. I know.